---
title: "Modification occurence in PUS7_KD SARS-CoV-2"
date: "`r Sys.Date()`"
output:
  rmdformats::html_clean:
    code_folding: hide
    highlight: default 
    lightbox: true
    thumbnails: false
---


```{r setup, echo=FALSE, cache=FALSE, include=F}
library(knitr)
library(rmdformats)
library(rmarkdown)
library(tidyverse)
library(ggpubr)
library(GGally)
library(tidyr)
library(R.utils)
library(reshape2)

## Global options
options(max.print="75")
opts_chunk$set(echo=FALSE,
               cache=TRUE,
               prompt=FALSE,
               tidy=FALSE,
               comment=NA,
               message=FALSE,
               warning=FALSE,
               dev=c("png","pdf"))
opts_knit$set(width=75)

ROOTDIR="/Volumes/scratch/TSSM/cugolini/cov"

# Function the returns full path from basedir
bdp <- function(relpath){
        return(paste0(ROOTDIR,"/",relpath))
}


fancy_scientific <- function(l) {
     # turn in to character string in scientific notation
     l <- format(l, scientific = TRUE)
     # quote the part before the exponent to keep all the digits
     l <- gsub("^(.*)e", "'\\1'e", l)
     # turn the 'e+' into plotmath format
     l <- gsub("e", "%*%10^", l)
     # return this as an expression
     parse(text=l)
}


vlab <- theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust=0.5))

### GENERAL DATA 

tx <- read_tsv(bdp("analysis/recappable_assembly/two_datasets/assemblies/pinfish/consensus_extraction/orf_annotate/orf_annotate.bed"), col_names=c("chr", "start", "end", "name", "score", "strand", "cdsStart", "cdsEnd", ".", "ex", "exLen", "exSt"), col_types="cnncncnnnncc")

names <- select(tx, orig=name) %>% separate(orig, into=c("id", "protein"), sep="#", remove=F) %>%
  mutate(protein=case_when(is.na(protein)~"Unknown", T~protein)) %>%
  separate(protein, into=c("sp", "uniprot_id", "protein"), sep="\\|", remove=F) %>%
  select(-sp) %>%
  mutate(name=gsub("([^\\(]+).+", "\\1", protein),
         tip=as.numeric(gsub("([^\\(]+)\\(([^%]+)%/([^%]+)%\\)", "\\2", protein)),
         qip=as.numeric(gsub("([^\\(]+)\\(([^%]+)%/([^%]+)%\\)", "\\3", protein))) %>%
  select(-protein) %>%
  select(-orig) %>%
  mutate(name=case_when(is.na(name)~"Unknown", T~name))

names$name[names$id == "efad7b96-ac2e-4ce1-9b83-863ffdb18eac|116::NC_045512v2:11-29873"] <- "ORF10_SARS2"
names$name[names$id == "de81ef19-655d-4ced-a9cc-cb8384001058|107::NC_045512v2:11-29874"] <- "ORF9D_SARS2"

tx_lengths <-read.table(bdp("analysis/recappable_assembly/two_datasets/assemblies/pinfish/aln_consensus.bed"), sep = '\t',header = FALSE) %>%
  separate(V11, into=c("ex1","ex2","ex3") ,sep=",", remove=F) 
tx_lengths[is.na(tx_lengths)] <- 0
tx_lengths <- as.data.frame(tx_lengths) %>% mutate(sumrow= as.numeric(ex1)  + as.numeric(ex2)+as.numeric(ex3))

#### blacklist IVT junctions

sitelist <- read_tsv(bdp("scripts_new/backupped_data/sites_blacklist.txt")) 
colnames(sitelist) <- c("ref_kmer","Modification type","modif")

IVT <-read_tsv(bdp("scripts_new/backupped_data/IVT_junctions.bed"), col_names = c("chrom","start","end","id","score","strand"))
junction_sites <- IVT %>% select(start,end) %>% unlist(use.names=FALSE) %>% sort()
blacklist_IVT <- vector()
for (i in junction_sites){
  temp <- seq(from = i-25, to = i+25, by = 1)
  blacklist_IVT <- c(blacklist_IVT,temp)
}
blacklist_IVT <- blacklist_IVT[which(blacklist_IVT > 0)]

#### blacklist exon junctions

assembly <-read.table(bdp("scripts_new/backupped_data/data_huxelerate_extraction/transcriptome_assembly/aln_consensus.bed"), col.names = c("chrom","start","end","id","score","strand","thickStart","thickEnd","itemRgb","blockCount","blockSizes","blockStarts"), sep="\t")
assembly_junction_sites <- assembly %>% select(start,end,id,blockSizes,blockStarts) %>% 
  separate(blockSizes, into = c("blSize1","blSize2","blSize3"), sep = ",") %>%
  separate(blockStarts, into = c("blStart1","blStart2","blStart3"), sep = ",") %>%
  mutate(blStart1 = (as.numeric(blStart1) + as.numeric(start))) %>%
  mutate(blStart2 = (as.numeric(blStart2) + as.numeric(start))) %>%
  mutate(blStart3 = (as.numeric(blStart3) + as.numeric(start))) %>%
  mutate(blEnd1 = (as.numeric(blStart1) + as.numeric(blSize1))) %>%
  mutate(blEnd2 = (as.numeric(blStart2) + as.numeric(blSize2))) %>%
  mutate(blEnd3 = (as.numeric(blStart3) + as.numeric(blSize3))) 

#canonicity
canonicity <- select(assembly,start,end,id) %>%
  mutate(canonicity=ifelse(start>100,"NC", "C")) %>%
  mutate(canonicity=ifelse(end<29000,"NC", canonicity)) %>%
  select(-start,-end)
colnames(canonicity) <- c("ref_id","canonicity") 


#### functions

pvalues_fisher_method = function(pvalues){
  keep = (pvalues >= 0) & (pvalues <= 1)
  pvalues[pvalues == 0] = 1e-285
  lnp = log(pvalues)
  chisq = (-2) * rowSums(lnp)
  df = 2 * length(lnp)
  fisher_pval = stats::pchisq(chisq, df, lower.tail=FALSE)
  return(fisher_pval)
}

junc_blacklist = function(tx) {
  corr_row <- subset(assembly_junction_sites, assembly_junction_sites$id %in% tx)
  blacklist_vec <- as.numeric(as.vector(corr_row[1,7:12]))
  blacklist_vec <- blacklist_vec[!is.na(blacklist_vec)]
  black_int <- vector()
  for (i in blacklist_vec){
    temp <- seq(from = i-15, to = i+15, by = 1)
    black_int <- c(black_int,temp)
  }
  black_int <- black_int[which(black_int > 0)] %>% sort()
  return(black_int)
}           

get_tx_length = function(ref_id) {
  tx_length = tx_lengths[with(tx_lengths, tx_lengths$V4 %in% ref_id),]$sumrow
  return(tx_length)
}

get_ORF = function(ref_id) {
  temp <- names[with(names, id %in% ref_id),]$name
  return(temp)
}


### LOAD DATA

tx_list <- list.files(path = "/Volumes/scratch/FN/TL/cugolini/cov/analysis/PUS7_KD_C37/map_to_recap_assembly/NANOCOMPORE/sampcomp/PUS7_KD_WT/",pattern = "*_results.tsv" , full.names = TRUE,  recursive = T)

PUS7_KD_WT <- lapply(tx_list, function(x){
  x <- read_tsv(x, col_types="ncccccnnncncn")
  x$SAMPLEID <- (separate(data.frame(A = x$cluster_counts), col = "A" , into = c("X", "Y","Z"), sep = "_(?=[0-9])"))$X
  x <- x[!x$SAMPLEID == "NC", ]
  x <- x %>% mutate(ref_kmer=gsub("T","U", ref_kmer))
})

```


### For every transcript of the assembly, a plot for PUS7_KD and WT modifications 
```{r, fig.width=18, fig.height=14}

PUS7_KD_WT <- PUS7_KD_WT[sapply(PUS7_KD_WT, function(x) dim(x)[1]) > 0] #delete 0-length dataframes
total <- bind_rows(PUS7_KD_WT) %>%
  rowwise() %>%
  mutate(ORF=get_ORF(ref_id)) %>%
  mutate(ref_id = (separate(data.frame(A = ref_id), col = "A" , into = c("X", "Y"), sep = "::"))$X) %>%
  rowwise() %>%
  mutate(tx_length=get_tx_length(ref_id)) %>%
  mutate(start_end_sites=ifelse(pos< 100, "pstart", NA)) %>%
  mutate(start_end_sites=ifelse(pos>(tx_length-40), "pend",start_end_sites )) %>%
  left_join(sitelist,by="ref_kmer") %>%
  dplyr::rename(IUPAC=`Modification type`) %>%
  mutate(IVT = ifelse(as.integer(genomicPos) %in% blacklist_IVT, "IVT junction", "No junction"))
total$IUPAC[is.na(total$IUPAC)] <- "Others"
total_split <- split(total,total$ref_id)


lapply(total_split,function(x){
  blacklist_junctions <- junc_blacklist(unique(x$ref_id))
  x %>% 
    mutate(IVT = ifelse(as.integer(genomicPos) %in% blacklist_junctions, "ORF junction", IVT))%>%
    {
    ggplot(., aes(x=abs(Logit_LOR), y=-log10(GMM_logit_pvalue),  color=IUPAC)) +
      geom_point(size=2) +
      scale_color_manual(values=c("#FF0000","#999999", "#56B4E9")) +
      ggtitle(unique(x$ORF), subtitle = unique(x$ref_id)) +
      theme_bw(28)
  }
})


```


### UGUAR grouped per fragments

```{r, fig.width=18, fig.height=14}

tx_list <- list.files(path = "/Volumes/scratch/FN/TL/cugolini/cov/analysis/PUS7_KD_C37/map_to_recap_assembly/NANOCOMPORE/sampcomp/PUS7_KD_WT/",pattern = "*_results.tsv" , full.names = TRUE,  recursive = T)

fragments <- read_tsv(bdp("scripts_new/backupped_data/RAPID/fragments_genomic_coord_UCSC.txt"))

pvalues_fisher_method = function(pvalues){
  keep = (pvalues >= 0) & (pvalues <= 1)
  pvalues[pvalues == 0] = 1e-285
  lnp = log(pvalues)
  chisq = (-2) * rowSums(lnp)
  df = 2 * length(lnp)
  fisher_pval = stats::pchisq(chisq, df, lower.tail=FALSE)
  return(fisher_pval)
}

junc_blacklist = function(tx) {
  corr_row <- subset(assembly_junction_sites, assembly_junction_sites$id %in% tx)
  blacklist_vec <- as.numeric(as.vector(corr_row[1,7:12]))
  blacklist_vec <- blacklist_vec[!is.na(blacklist_vec)]
  black_int <- vector()
  for (i in blacklist_vec){
    temp <- seq(from = i-15, to = i+15, by = 1)
    black_int <- c(black_int,temp)
  }
  black_int <- black_int[which(black_int > 0)] %>% sort()
  return(black_int)
}

assign_fragment = function(df) {
  df$fragment_ID <- "No_Fragment"
  df$genomicPos <- as.numeric(df$genomicPos)
  for (i in 1:length(fragments$ID)){
    df$fragment_ID <- ifelse(df$genomicPos >= fragments$start[i] & df$genomicPos <= fragments$end[i], fragments$ID[i], df$fragment_ID)
  }
  return(df)
}


total <- read_tsv(tx_list, col_types="ncccccnnncncn") %>%
  subset(cluster_counts!="NC") %>%
  mutate(SAMPLEID = (separate(data.frame(A = cluster_counts), col = "A" , into = c("X", "Y","Z"), sep = "_(?=[0-9])"))$X) %>%
  mutate(ref_kmer=gsub("T","U", ref_kmer))

tmp <- split(total, total$genomicPos)
toplot <- lapply(X = tmp,FUN = function(y){
  pvalues <- y %>%
    select(ref_id,GMM_logit_pvalue) %>%
    spread(ref_id,GMM_logit_pvalue)
  txcov <- length(y$ref_id)
  y <- y %>%
    dplyr::slice(1:1) %>%
    mutate(tx_cov=txcov) %>%
    mutate(GMM_logit_pvalue = ifelse(pvalues_fisher_method(pvalues) == 0, 1e-285,pvalues_fisher_method(pvalues)))
  return(y)
})

final <- bind_rows(toplot)

final <- final %>%
  mutate(IVT = ifelse(as.integer(final$genomicPos) %in% blacklist_IVT, "IVT junction", "No junction")) %>%
  mutate(ref_id = sub("\\::.*", "",ref_id))

tmp <- final
for(i in length(tmp$ref_id)){
  bl_temp <- junc_blacklist(tmp$ref_id[i])
  tmp$IVT[i] <- ifelse(as.integer(tmp$genomicPos[i]) %in% bl_temp, "ORF junction", tmp$IVT[i])
}
final <- tmp %>%
  left_join(sitelist)
final$`Modification type`[is.na(final$`Modification type`)] <- "Others"

paper_sites <- c(22322,23317,27164,28417,28759,28927,29418)

final <- final %>%
  mutate(paper = case_when(
    genomicPos %in% paper_sites ~ "paper"
    ))

paper_sites_table <- final %>% 
  subset(paper == "paper") %>%
  select(genomicPos,ref_kmer,GMM_logit_pvalue,KS_dwell_pvalue,KS_intensity_pvalue,Logit_LOR,SAMPLEID,tx_cov,IVT,`Modification type`) 
j1 <- c("GMM_logit_pvalue","KS_dwell_pvalue","KS_intensity_pvalue","Logit_LOR")
paper_sites_table[j1] <- lapply(paper_sites_table[j1], format, digits = 4)

write.table(paper_sites_table, "/Users/camillaugolini/Desktop/paper_sites.txt", sep="\t",quote = F, row.names = F, col.names = T)

final_UGUAR <- final %>%
  subset(final$`Modification type`== "UGUAR")
final_UGUAR <- assign_fragment(final_UGUAR)


final_UGUAR <- final_UGUAR %>%
  mutate(tx_cov=factor(tx_cov))
final_UGUAR$fragment_ID<-factor(final_UGUAR$fragment_ID,levels=c("No_Fragment","Fragment1_5UTR","Fragment1_2","Fragment2_3","Fragment3_4","Fragment4_5","Fragment5","Fragment6","Fragment6_7","Fragment7_8","Fragment8_9","Fragment9_10","Fragment10_3UTR"))

bed_UGUAR_fragments <- final_UGUAR %>%
  subset(fragment_ID!="No_Fragment") %>%
  select(chr,genomicPos,fragment_ID,strand)%>%
  mutate(genomicPos_end=(genomicPos+5),genomicPos_start=(genomicPos-5),score=0)%>%
  mutate(fragment_ID=paste0(fragment_ID,"_",genomicPos))%>%
  select(-genomicPos)%>%
  relocate(chr,genomicPos_start,genomicPos_end,fragment_ID,score,strand)

write.table(bed_UGUAR_fragments,"/Volumes/scratch/FN/TL/cugolini/cov/analysis/PUS7_KD_C37/map_to_genome/pileup/regions_UGUAR_fragments.bed",sep = "\t",quote = F,row.names = F,col.names = F)



p <- ggplot(final_UGUAR, aes(x=genomicPos, y=-log10(GMM_logit_pvalue), size=tx_cov, color=fragment_ID,shape=IVT)) +
      geom_point() +
      scale_color_manual(breaks = c("No_Fragment","Fragment1_5UTR","Fragment1_2","Fragment2_3","Fragment3_4","Fragment4_5","Fragment5","Fragment6","Fragment6_7","Fragment7_8","Fragment8_9","Fragment9_10","Fragment10_3UTR"),values=c("black","blue", "green","grey","gold","coral","aquamarine","darkgreen","navy","deeppink","magenta","cyan","orange")) +
      scale_size_discrete(name = "N° of transcripts") +
      scale_shape_manual(values=c(3, 16))+
      #ggrepel::geom_label_repel(data=filter(x, -log10(GMM_logit_pvalue)>=5), aes(label=paste0(ref_kmer, " (",genomicPos,")")), colour="black", size=5) +
      ggtitle("UGUAR occurrence over all transcripts ") +
      theme_bw(28) +
      theme(plot.title = element_text(size = 30, face = "bold"),
            plot.subtitle = element_text(size = 28),
            legend.title=element_text(size=26),
            legend.text=element_text(size=24))
p

ggsave(p,file="/Volumes/scratch/FN/TL/cugolini/cov/scripts/downstream_PUS7KD/PUS7KD_C37/UGUAR.pdf", width = 15,height = 10)


final <- assign_fragment(final)
final <- final %>%
  mutate(tx_cov=factor(tx_cov)) %>%
  mutate(IVT=factor(IVT)) %>%
  mutate(`Modification type`=factor(`Modification type`)) 


p <- ggplot(final, aes(x=genomicPos, y=-log10(GMM_logit_pvalue), size=tx_cov, color=`Modification type`, shape=IVT)) +
      geom_point() +
      scale_color_manual(breaks =c("Others","DRACH","UGUAR"), values=c("black","blue","deeppink")) +
      scale_size_discrete(name = "N° of transcripts") +
      scale_shape_manual(values=c(3, 16, 14))+
      ggrepel::geom_label_repel(data=filter(final, -log10(GMM_logit_pvalue)>=5), aes(label=paste0(ref_kmer, " (",genomicPos,")")), colour="black", size=5) +
      ggtitle("Modifications occurrence over all transcripts  ") +
      theme_bw(28) +
    theme(plot.title = element_text(size = 30, face = "bold"),
            plot.subtitle = element_text(size = 28),
            legend.title=element_text(size=26),
            legend.text=element_text(size=24))
p
ggsave(p,file="/Volumes/scratch/FN/TL/cugolini/cov/scripts/downstream_PUS7KD/PUS7KD_C37/all_mods.pdf", width = 20,height = 20)

```

### plots IVT vs WT  and IVT vs PUS7KD
```{r, fig.width=18, fig.height=14}

tx_list <- list.files(path = "/Volumes/scratch/FN/TL/cugolini/cov/analysis/PUS7_KD_C37/map_to_recap_assembly/NANOCOMPORE/sampcomp/WT_IVT",pattern = "*_results.tsv" , full.names = TRUE,  recursive = T)

fragments <- read_tsv(bdp("scripts_new/backupped_data/RAPID/fragments_genomic_coord_UCSC.txt"))

assign_fragment = function(df) {
  df$fragment_ID <- "No_Fragment"
  df$genomicPos <- as.numeric(df$genomicPos)
  for (i in 1:length(fragments$ID)){
    df$fragment_ID <- ifelse(df$genomicPos >= fragments$start[i] & df$genomicPos <= fragments$end[i], fragments$ID[i], df$fragment_ID)
  }
  return(df)
}


datasets <- lapply(tx_list, function(x){
  x <- read_tsv(x, col_types="ncccccnnncncn")
  x$SAMPLEID <- (separate(data.frame(A = x$cluster_counts), col = "A" , into = c("X", "Y","Z"), sep = "_(?=[0-9])"))$X
  x <- x[!x$SAMPLEID == "NC", ]
  x <- x %>% mutate(ref_kmer=gsub("T","U", ref_kmer))
})

datasets <- datasets[sapply(datasets, function(x) dim(x)[1]) > 0] #delete 0-length dataframes
total <- bind_rows(datasets) %>%
  rowwise() %>%
  mutate(ORF=get_ORF(ref_id)) %>%
  mutate(ref_id = (separate(data.frame(A = ref_id), col = "A" , into = c("X", "Y"), sep = "::"))$X) %>%
  rowwise() %>%
  mutate(tx_length=get_tx_length(ref_id)) %>%
  mutate(start_end_sites=ifelse(pos< 100, "pstart", NA)) %>%
  mutate(start_end_sites=ifelse(pos>(tx_length-40), "pend",start_end_sites )) %>%
  left_join(sitelist,by="ref_kmer") %>%
  dplyr::rename(IUPAC=`Modification type`)
total$IUPAC[is.na(total$IUPAC)] <- "Others"

tmp <- split(total, total$genomicPos)
toplot <- lapply(X = tmp,FUN = function(y){
  pvalues <- y %>%
    select(ref_id,GMM_logit_pvalue) %>%
    spread(ref_id,GMM_logit_pvalue)
  txcov <- length(y$ref_id)
  y <- y %>%
    dplyr::slice(1:1) %>%
    mutate(tx_cov=txcov) %>%
    mutate(GMM_logit_pvalue = ifelse(pvalues_fisher_method(pvalues) == 0, 1e-285,pvalues_fisher_method(pvalues)))
  return(y)
})

final <- bind_rows(toplot)
final <- final %>%
  mutate(IVT = ifelse(as.integer(genomicPos) %in% blacklist_IVT, "IVT junction", "No junction"))

tmp <- final
for(i in length(tmp$ref_id)){
  bl_temp <- junc_blacklist(tmp$ref_id[i])
  tmp$IVT[i] <- ifelse(as.integer(tmp$genomicPos[i]) %in% bl_temp, "ORF junction", tmp$IVT[i])
}
final <- tmp %>%
  left_join(sitelist)
final$`Modification type`[is.na(final$`Modification type`)] <- "Others"

final_UGUAR <- final %>%
  subset(final$`Modification type`== "UGUAR")
final_UGUAR <- assign_fragment(final_UGUAR)


final_UGUAR <- final_UGUAR %>%
  mutate(tx_cov=factor(tx_cov))
final_UGUAR$fragment_ID<-factor(final_UGUAR$fragment_ID,levels=c("No_Fragment","Fragment1_5UTR","Fragment1_2","Fragment2_3","Fragment3_4","Fragment4_5","Fragment5","Fragment6","Fragment6_7","Fragment7_8","Fragment8_9","Fragment9_10","Fragment10_3UTR"))

p <- ggplot(final_UGUAR, aes(x=genomicPos, y=-log10(GMM_logit_pvalue), size=tx_cov, color=fragment_ID,shape=IVT)) +
      geom_point() +
      scale_color_manual(breaks = c("No_Fragment","Fragment1_5UTR","Fragment1_2","Fragment2_3","Fragment3_4","Fragment4_5","Fragment5","Fragment6","Fragment6_7","Fragment7_8","Fragment8_9","Fragment9_10","Fragment10_3UTR"),values=c("black","blue", "green","grey","gold","coral","aquamarine","darkgreen","navy","deeppink","magenta","cyan","orange")) +
      scale_size_discrete(name = "N° of transcripts") +
      scale_shape_manual(values=c(3, 16))+
      #ggrepel::geom_label_repel(data=filter(x, -log10(GMM_logit_pvalue)>=5), aes(label=paste0(ref_kmer, " (",genomicPos,")")), colour="black", size=5) +
      ggtitle("UGUAR occurrence over all transcripts ") +
      theme_bw(28) +
      theme(plot.title = element_text(size = 30, face = "bold"),
            plot.subtitle = element_text(size = 28),
            legend.title=element_text(size=26),
            legend.text=element_text(size=24))
p


ggsave(p,file="/Volumes/scratch/FN/TL/cugolini/cov/scripts/downstream_PUS7KD/PUS7KD_C37/UGUAR_IVTvsWT.pdf", width = 20,height = 20)



tx_list <- list.files(path = "/Volumes/scratch/FN/TL/cugolini/cov/analysis/PUS7_KD_C37/map_to_recap_assembly/NANOCOMPORE/sampcomp/PUS7_KD_IVT",pattern = "*_results.tsv" , full.names = TRUE,  recursive = T)

datasets <- lapply(tx_list, function(x){
  x <- read_tsv(x, col_types="ncccccnnncncn")
  x$SAMPLEID <- (separate(data.frame(A = x$cluster_counts), col = "A" , into = c("X", "Y","Z"), sep = "_(?=[0-9])"))$X
  x <- x[!x$SAMPLEID == "NC", ]
  x <- x %>% mutate(ref_kmer=gsub("T","U", ref_kmer))
})

datasets <- datasets[sapply(datasets, function(x) dim(x)[1]) > 0] #delete 0-length dataframes
total <- bind_rows(datasets) %>%
  rowwise() %>%
  mutate(ORF=get_ORF(ref_id)) %>%
  mutate(ref_id = (separate(data.frame(A = ref_id), col = "A" , into = c("X", "Y"), sep = "::"))$X) %>%
  rowwise() %>%
  mutate(tx_length=get_tx_length(ref_id)) %>%
  mutate(start_end_sites=ifelse(pos< 100, "pstart", NA)) %>%
  mutate(start_end_sites=ifelse(pos>(tx_length-40), "pend",start_end_sites )) %>%
  left_join(sitelist,by="ref_kmer") %>%
  dplyr::rename(IUPAC=`Modification type`)
total$IUPAC[is.na(total$IUPAC)] <- "Others"

tmp <- split(total, total$genomicPos)
toplot <- lapply(X = tmp,FUN = function(y){
  pvalues <- y %>%
    select(ref_id,GMM_logit_pvalue) %>%
    spread(ref_id,GMM_logit_pvalue)
  txcov <- length(y$ref_id)
  y <- y %>%
    dplyr::slice(1:1) %>%
    mutate(tx_cov=txcov) %>%
    mutate(GMM_logit_pvalue = ifelse(pvalues_fisher_method(pvalues) == 0, 1e-285,pvalues_fisher_method(pvalues)))
  return(y)
})

final <- bind_rows(toplot)

final <- final %>%
  mutate(IVT = ifelse(as.integer(genomicPos) %in% blacklist_IVT, "IVT junction", "No junction"))

tmp <- final
for(i in length(tmp$ref_id)){
  bl_temp <- junc_blacklist(tmp$ref_id[i])
  tmp$IVT[i] <- ifelse(as.integer(tmp$genomicPos[i]) %in% bl_temp, "ORF junction", tmp$IVT[i])
}
final <- tmp %>%
  left_join(sitelist)
final$`Modification type`[is.na(final$`Modification type`)] <- "Others"

final_UGUAR <- final %>%
  subset(final$`Modification type`== "UGUAR")
final_UGUAR <- assign_fragment(final_UGUAR)


final_UGUAR <- final_UGUAR %>%
  mutate(tx_cov=factor(tx_cov))
final_UGUAR$fragment_ID<-factor(final_UGUAR$fragment_ID,levels=c("No_Fragment","Fragment1_5UTR","Fragment1_2","Fragment2_3","Fragment3_4","Fragment4_5","Fragment5","Fragment6","Fragment6_7","Fragment7_8","Fragment8_9","Fragment9_10","Fragment10_3UTR"))

p <- ggplot(final_UGUAR, aes(x=genomicPos, y=-log10(GMM_logit_pvalue), size=tx_cov, color=fragment_ID,shape=IVT)) +
      geom_point() +
      scale_color_manual(breaks = c("No_Fragment","Fragment1_5UTR","Fragment1_2","Fragment2_3","Fragment3_4","Fragment4_5","Fragment5","Fragment6","Fragment6_7","Fragment7_8","Fragment8_9","Fragment9_10","Fragment10_3UTR"),values=c("black","blue", "green","grey","gold","coral","aquamarine","darkgreen","navy","deeppink","magenta","cyan","orange")) +
      scale_size_discrete(name = "N° of transcripts") +
      scale_shape_manual(values=c(3, 16))+
      #ggrepel::geom_label_repel(data=filter(x, -log10(GMM_logit_pvalue)>=5), aes(label=paste0(ref_kmer, " (",genomicPos,")")), colour="black", size=5) +
      ggtitle("UGUAR occurrence over all transcripts ") +
      theme_bw(28) +
      theme(plot.title = element_text(size = 30, face = "bold"),
            plot.subtitle = element_text(size = 28),
            legend.title=element_text(size=26),
            legend.text=element_text(size=24))
p

ggsave(p,file="/Volumes/scratch/FN/TL/cugolini/cov/scripts/downstream_PUS7KD/PUS7KD_C37/UGUAR_IVTvsPUS7KD.pdf", width = 20,height = 20)


```

### For every transcript of the assembly, a plot for PUS7_KD and WT modifications 
```{r, fig.width=18, fig.height=14}

#### functions

pvalues_fisher_method = function(pvalues){
  keep = (pvalues >= 0) & (pvalues <= 1)
  pvalues[pvalues == 0] = 1e-285
  lnp = log(pvalues)
  chisq = (-2) * rowSums(lnp)
  df = 2 * length(lnp)
  fisher_pval = stats::pchisq(chisq, df, lower.tail=FALSE)
  return(fisher_pval)
}

junc_blacklist = function(tx) {
  corr_row <- subset(assembly_junction_sites, assembly_junction_sites$id %in% tx)
  blacklist_vec <- as.numeric(as.vector(corr_row[1,7:12]))
  blacklist_vec <- blacklist_vec[!is.na(blacklist_vec)]
  black_int <- vector()
  for (i in blacklist_vec){
    temp <- seq(from = i-15, to = i+15, by = 1)
    black_int <- c(black_int,temp)
  }
  black_int <- black_int[which(black_int > 0)] %>% sort()
  return(black_int)
}           

get_tx_length = function(ref_id) {
  tx_length = tx_lengths[with(tx_lengths, tx_lengths$V4 %in% ref_id),]$sumrow
  return(tx_length)
}

get_ORF = function(ref_id) {
  temp <- names[with(names, id %in% ref_id),]$name
  return(temp)
}

tx_list <- list.files(path = "/Volumes/scratch/FN/TL/cugolini/cov/analysis/PUS7_KD_C37/map_to_recap_assembly/NANOCOMPORE/sampcomp/PUS7_KD_IVT",pattern = "*_results.tsv" , full.names = TRUE,  recursive = T)

datasets <- lapply(tx_list, function(x){
  x <- read_tsv(x, col_types="ncccccnnncncn")
  x$SAMPLEID <- (separate(data.frame(A = x$cluster_counts), col = "A" , into = c("X", "Y","Z"), sep = "_(?=[0-9])"))$X
  x <- x[!x$SAMPLEID == "NC", ]
  x <- x %>% mutate(ref_kmer=gsub("T","U", ref_kmer))
})

datasets <- datasets[sapply(datasets, function(x) dim(x)[1]) > 0] #delete 0-length dataframes
total <- bind_rows(datasets) %>%
  rowwise() %>%
  mutate(ORF=get_ORF(ref_id)) %>%
  mutate(ref_id = (separate(data.frame(A = ref_id), col = "A" , into = c("X", "Y"), sep = "::"))$X) %>%
  rowwise() %>%
  mutate(tx_length=get_tx_length(ref_id)) %>%
  mutate(start_end_sites=ifelse(pos< 100, "pstart", NA)) %>%
  mutate(start_end_sites=ifelse(pos>(tx_length-40), "pend",start_end_sites )) %>%
  left_join(sitelist,by="ref_kmer") %>%
  dplyr::rename(IUPAC=`Modification type`)
total$IUPAC[is.na(total$IUPAC)] <- "Others"
total <- subset(total, !(genomicPos %in% blacklist_IVT))
blacklist_junctions <- junc_blacklist(unique(total$ref_id))
total <- subset(total, !(genomicPos %in% blacklist_junctions))

tx_list <- list.files(path = "/Volumes/scratch/FN/TL/cugolini/cov/analysis/PUS7_KD_C37/map_to_recap_assembly/NANOCOMPORE/sampcomp/WT_IVT",pattern = "*_results.tsv" , full.names = TRUE,  recursive = T)

datasets <- lapply(tx_list, function(x){
  x <- read_tsv(x, col_types="ncccccnnncncn")
  x$SAMPLEID <- (separate(data.frame(A = x$cluster_counts), col = "A" , into = c("X", "Y","Z"), sep = "_(?=[0-9])"))$X
  x <- x[!x$SAMPLEID == "NC", ]
  x <- x %>% mutate(ref_kmer=gsub("T","U", ref_kmer))
})

datasets <- datasets[sapply(datasets, function(x) dim(x)[1]) > 0] #delete 0-length dataframes
total_WT <- bind_rows(datasets) %>%
  rowwise() %>%
  mutate(ORF=get_ORF(ref_id)) %>%
  mutate(ref_id = (separate(data.frame(A = ref_id), col = "A" , into = c("X", "Y"), sep = "::"))$X) %>%
  rowwise() %>%
  mutate(tx_length=get_tx_length(ref_id)) %>%
  mutate(start_end_sites=ifelse(pos< 100, "pstart", NA)) %>%
  mutate(start_end_sites=ifelse(pos>(tx_length-40), "pend",start_end_sites )) %>%
  left_join(sitelist,by="ref_kmer") %>%
  dplyr::rename(IUPAC=`Modification type`)
total_WT$IUPAC[is.na(total_WT$IUPAC)] <- "Others"
total_WT <- subset(total_WT, !(genomicPos %in% blacklist_IVT))
total_WT <- total_WT %>%
  mutate(IVT = ifelse(as.integer(genomicPos) %in% blacklist_IVT, "IVT junction", "No junction"))
for(i in length(total_WT$ref_id)){
  bl_temp <- junc_blacklist(total_WT$ref_id[i])
  total_WT$IVT[i] <- ifelse(as.integer(total_WT$genomicPos[i]) %in% bl_temp, "ORF junction", total_WT$IVT[i])
}



total_WT_PUS7KD <- rbind(total,total_WT)
total_UGUAR <- total_WT_PUS7KD %>%
  subset(total_WT_PUS7KD$IUPAC== "UGUAR")
total_UGUAR <- assign_fragment(total_UGUAR)
total_split <- split(total_UGUAR,total_UGUAR$ref_id)


pdf(file = "/Volumes/scratch/FN/TL/cugolini/cov/scripts/downstream_PUS7KD/PUS7KD_C37/UGUAR_IVTvsPUS7KD_WT_per_transcript.pdf",width =10,height =10)
lapply(total_split,function(x){
  x$fragment_ID<-factor(x$fragment_ID,levels=c("No_Fragment","Fragment1_5UTR","Fragment1_2","Fragment2_3","Fragment3_4","Fragment4_5","Fragment5","Fragment6","Fragment6_7","Fragment7_8","Fragment8_9","Fragment9_10","Fragment10_3UTR"))
  x %>%{
    ggplot(., aes(x=abs(Logit_LOR), y=-log10(GMM_logit_pvalue), color=fragment_ID,shape=SAMPLEID)) +
      geom_point(size=2) +
      scale_color_manual(breaks = c("No_Fragment","Fragment1_5UTR","Fragment1_2","Fragment2_3","Fragment3_4","Fragment4_5","Fragment5","Fragment6","Fragment6_7","Fragment7_8","Fragment8_9","Fragment9_10","Fragment10_3UTR"),values=c("black","blue", "green","grey","gold","coral","aquamarine","darkgreen","navy","deeppink","magenta","cyan","orange")) +
      scale_shape_manual(values=c(3, 16))+
      ggrepel::geom_label_repel(data=filter(., fragment_ID!="No_Fragment" ) ,aes(label=paste0(ref_kmer, " (",genomicPos,")")), colour="black", size=4) +
      ggtitle(unique(x$ORF), subtitle = unique(x$ref_id)) +
      theme_bw(18)
  }
})
dev.off()



## calculate tables for p-value separation between WT vs IVT and KD vs IVT per transcript model

total_UGUAR <- total %>% subset(IUPAC== "UGUAR")
total_WT_UGUAR <-total_WT %>% subset(IUPAC== "UGUAR")

total_WT_PUS7KD <- full_join(total_UGUAR,total_WT_UGUAR,by=c("pos","chr","ref_id","strand","ref_kmer","ORF","tx_length","IUPAC")) %>%
  subset(IUPAC== "UGUAR")
total_WT_PUS7KD <- assign_fragment(total_WT_PUS7KD)
total_WT_PUS7KD$GMM_logit_pvalue.y[is.na(total_WT_PUS7KD$GMM_logit_pvalue.y)] <- 1
total_WT_PUS7KD$GMM_logit_pvalue.x[is.na(total_WT_PUS7KD$GMM_logit_pvalue.x)] <- 1
total_WT_PUS7KD <- total_WT_PUS7KD %>%
  mutate(GMMpvalue_WToverPUS7KD=GMM_logit_pvalue.y/GMM_logit_pvalue.x) %>%
  left_join(canonicity,by="ref_id") %>%
  mutate(pval_category=ifelse(-log10(abs(GMMpvalue_WToverPUS7KD))<=1,"identical","non identical")) %>%
  mutate(pval_category=ifelse(-log10(abs(GMMpvalue_WToverPUS7KD))>1 & -log10(abs(GMMpvalue_WToverPUS7KD))<=3 ,"small diff",pval_category)) %>%
  mutate(pval_category=ifelse(-log10(abs(GMMpvalue_WToverPUS7KD))>3 & -log10(abs(GMMpvalue_WToverPUS7KD))<=10 ,"diff",pval_category)) %>%
  mutate(pval_category=ifelse(-log10(abs(GMMpvalue_WToverPUS7KD))>10 & -log10(abs(GMMpvalue_WToverPUS7KD))<=100 ,"large diff",pval_category)) 
  
total_split <- split(total_WT_PUS7KD,total_WT_PUS7KD$ref_id)


plot_list <- lapply(total_split,function(x){
  x %>% ggplot(., aes(x=fragment_ID, y=GMMpvalue_WToverPUS7KD)) +
    geom_boxplot() +
    theme_bw(8) +
    ggtitle(unique(x$ref_id), subtitle = paste0(unique(x$ORF),"-",unique(x$canonicity)))
})

p<- ggarrange(plotlist=plot_list)

ggsave(p,file="/Volumes/scratch/FN/TL/cugolini/cov/scripts/downstream_PUS7KD/PUS7KD_C37/wt_pus_pval.pdf", width = 20,height = 20)

```



